phoenicis-configuration/src/main/java/org/phoenicis/configuration/localisation/Localisation.java

```diff
@@ -76,11 +85,153 @@ public static String tr(String text, Object[] objects) {
 
     /**
      * Translates the given Translatable object, by creating a new instance of it with its corresponding translations
-     * @param translatable Translatable object
+     *
      * @return translated object
      */
-    public static <T extends Translatable<?>> T tr(Translatable<T> translatable) {
-        return translatable.translate();
+    public static <T> T tr(T translatable) {
+        if (translatable instanceof String) {
+            @SuppressWarnings("unchecked")
+            final T translateResult = (T) trString(translatable.toString());
+            return translateResult;
+        }
+
+        if (translatable instanceof List) {
+            @SuppressWarnings("unchecked")
+            final T translateResult = (T) trList((List) translatable);
+            return translateResult;
+        }
+
+        if (translatable instanceof Set) {
+            @SuppressWarnings("unchecked")
+            final T translateResult = (T) trSet((Set) translatable);
+            return translateResult;
+        }
+
+        final Class<?> clazz = translatable.getClass();
+        final Optional<Constructor<?>> translateCreator = findTranslateCreator(clazz);
+        final Optional<Class<?>> translateBuilder = findTranslateBuilder(clazz);
+
+        try {
+            if (translateCreator.isPresent()) {
+                return trFromTranslateCreator(translatable, translateCreator.get());
+            }
+
+            if (translateBuilder.isPresent()) {
+                return trFromTranslateBuilder(translatable, translateBuilder.get());
+            }
+
+        } catch (ReflectiveOperationException e) {
+            LOGGER.warn("Error while translating " + translatable + ". Falling back to the original value", e);
+        }
+
+        LOGGER.warn(translatable + " is neither a String, neither a collection, neither translatable. Ignoring");
+        return translatable;
+    }
+
+    /**
+     * Tells if a class is translatable or not
+     * @param clazz The class to test
+     * @return A boolean
+     */
+    public static boolean isTranslatable(Class<?> clazz) {
+        for (Annotation annotation : clazz.getAnnotations()) {
+            if (annotation.annotationType() == Translatable.class) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static List<?> trList(List<?> translatable) {
+        return translatable.stream().map(Localisation::tr).collect(Collectors.toList());
+    }
+
+    private static Set<?> trSet(Set<?> translatable) {
+        return translatable.stream().map(Localisation::tr).collect(Collectors.toSet());
+    }
+
+    private static <T> T trFromTranslateCreator(T translatable, Constructor<?> translateCreator)
```

Comment:
This is really strange because:
```java
T result = (T) translateCreator.newInstance(parameters.toArray());
```
shows, that the given constructor should be of type `Constructor<T>` or `Constructor<? extends T>`, right?